go/cmd/dolt/commands/cherry-pick.go

```diff
@@ -0,0 +1,466 @@
+func (cmd CherryPickCmd) EventType() eventsapi.ClientEventType {
+	return eventsapi.ClientEventType_CHERRY_PICK
+}
+
+// Exec executes the command
+func (cmd CherryPickCmd) Exec(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	ap := cmd.ArgParser()
+	help, usage := cli.HelpAndUsagePrinters(cli.GetCommandDocumentation(commandStr, dumpDocs, ap))
+	apr := cli.ParseArgsOrDie(ap, args, help)
+
+	// This command creates a commit, so we need user identity
+	if !cli.CheckUserNameAndEmail(dEnv) {
+		return 1
+	}
+
+	// TODO : support single commit cherry-pick only for now
+	if apr.NArg() == 0 {
+		usage()
+		return 1
+	} else if apr.NArg() > 1 {
+		return HandleVErrAndExitCode(errhand.BuildDError("multiple commits not supported yet.").SetPrintUsage().Build(), usage)
+	}
+
+	// check for clean working state
+	headRoot, err := dEnv.HeadRoot(ctx)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	workingRoot, err := dEnv.WorkingRoot(ctx)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	headHash, err := headRoot.HashOf()
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	workingHash, err := workingRoot.HashOf()
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	if !headHash.Equal(workingHash) {
+		return HandleVErrAndExitCode(errhand.BuildDError("You must commit any changes before using cherry-pick.").SetPrintUsage().Build(), usage)
+	}
+
+	cherryStr := apr.Arg(0)
+	if len(cherryStr) == 0 {
+		verr := errhand.BuildDError("error: cannot cherry-pick empty string").Build()
+		return HandleVErrAndExitCode(verr, usage)
+	}
+
+	newWorkingRoot, commitMsg, err := CherryPick(ctx, dEnv, workingRoot, cherryStr)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+
+	workingHash, err = newWorkingRoot.HashOf()
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+
+	if headHash.Equal(workingHash) {
+		cli.Println("No changes were made.")
+		return 0
+	}
+
+	err = dEnv.UpdateWorkingRoot(ctx, newWorkingRoot)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	res := AddCmd{}.Exec(ctx, "add", []string{"-A"}, dEnv)
+	if res != 0 {
+		return res
+	}
+
+	// Pass in the final parameters for the author string.
+	commitParams := []string{"-m", commitMsg}
+	authorStr, ok := apr.GetValue(cli.AuthorParam)
+	if ok {
+		commitParams = append(commitParams, "--author", authorStr)
+	}
+
+	return CommitCmd{}.Exec(ctx, "commit", commitParams, dEnv)
+}
+
+// CherryPick returns updated Root value for current HEAD and commit message of cherry-picked commit.
+func CherryPick(ctx context.Context, dEnv *env.DoltEnv, wRoot *doltdb.RootValue, cherryStr string) (*doltdb.RootValue, string, error) {
```

Comment:
(minor) Does this function need to be exported yet? Better to start off as package private unless we need it exported.